ADIS16477.cpp: 275:95: ERROR: statement has no effect

Build failed with the following error for ADIS16477 driver. I am using a custom sensor board with Raspberry Pi 3.

$ sudo make px4_rpi
[0/1] Re-running CMake...
-- PX4 version: v1.11.0-rc3-2014-g0d4f48a8d2
-- PX4 config file: /home/rusr/px4dev/Firmware/boards/px4/rpi/default.cmake
-- PX4 config: px4_rpi_default
-- PX4 platform: posix
-- PX4 lockstep: disabled
-- cmake build type: RelWithDebInfo
-- Building for code coverage
-- ccache enabled (export CCACHE_DISABLE=1 to disable)
-- build type is RelWithDebInfo
-- PX4 ECL: Very lightweight Estimation & Control Library v1.9.0-rc1-544-g4df0054
-- ROMFS: ROMFS/px4fmu_common
-- Configuring done
-- Generating done
-- Build files have been written to: /home/rusr/px4dev/Firmware/build/px4_rpi_default
[303/714] Building CXX object src/drivers/imu/a...les/drivers__imu__adis16477.dir/ADIS16477.cpp.o
FAILED: src/drivers/imu/adis16477/CMakeFiles/drivers__imu__adis16477.dir/ADIS16477.cpp.o 
/usr/bin/ccache /usr/bin/arm-linux-gnueabihf-g++  -DCONFIG_ARCH_BOARD_PX4_RPI -DMODULE_NAME=\"adis16477\" -DPX4_MAIN=adis16477_app_main -D__CUSTOM_FILE_IO__ -D__PX4_LINUX -D__PX4_POSIX -D__STDC_FORMAT_MACROS -Dnoreturn_function="__attribute__((noreturn))" -I../../boards/px4/rpi/src -I../../platforms/posix/src/px4/common/include -I. -Isrc/lib -I../../platforms/posix/src/px4/generic/generic/include -I../../platforms/common -I../../platforms/common/include -I../../src -I../../src/include -I../../src/lib -I../../src/lib/matrix -I../../src/modules -I../../platforms/posix/include -Iexternal/Install/include -O2 -g -DNDEBUG -fPIC   -g -fdata-sections -ffunction-sections -fomit-frame-pointer -fmerge-all-constants -fno-signed-zeros -fno-trapping-math -freciprocal-math -fno-math-errno -fno-strict-aliasing -fvisibility=hidden -include visibility.h -Wall -Wextra -Werror -Warray-bounds -Wcast-align -Wdisabled-optimization -Wdouble-promotion -Wfatal-errors -Wfloat-equal -Wformat-security -Winit-self -Wlogical-op -Wpointer-arith -Wshadow -Wuninitialized -Wunknown-pragmas -Wunused-variable -Wno-missing-field-initializers -Wno-missing-include-dirs -Wno-unused-parameter -fdiagnostics-color=always -fno-builtin-printf -fno-strength-reduce -Wformat=1 -Wunused-but-set-variable -Wno-format-truncation -fcheck-new -fno-exceptions -fno-rtti -fno-threadsafe-statics -Wreorder -Wno-overloaded-virtual -Wno-cast-align -std=gnu++14 -MD -MT src/drivers/imu/adis16477/CMakeFiles/drivers__imu__adis16477.dir/ADIS16477.cpp.o -MF src/drivers/imu/adis16477/CMakeFiles/drivers__imu__adis16477.dir/ADIS16477.cpp.o.d -o src/drivers/imu/adis16477/CMakeFiles/drivers__imu__adis16477.dir/ADIS16477.cpp.o -c ../../src/drivers/imu/adis16477/ADIS16477.cpp
../../src/drivers/imu/adis16477/ADIS16477.cpp: In member function ‘void ADIS16477::start()’:
../../src/drivers/imu/adis16477/ADIS16477.cpp:275:95: error: statement has no effect [-Werror=unused-value]
   px4_arch_gpiosetevent(_drdy_gpio, true, false, true, &ADIS16477::data_ready_interrupt, this);
                                                                                               ^
compilation terminated due to -Wfatal-errors.
cc1plus: all warnings being treated as errors
[308/714] Building CXX object src/drivers/magne...ter__st__lsm9ds1_mag.dir/lsm9ds1_mag_main.cpp.o
ninja: build stopped: subcommand failed.
Makefile:226: recipe for target 'px4_rpi' failed
make: *** [px4_rpi] Error 1

The issue is somewhat resolved by commenting out the -Werror in cmake/px4_add_common_flags.cmake.
Perhaps this is not a correct solution and the problem should be fixed in the master before closing this issue.

$ sudo make px4_rpi
[0/1] Re-running CMake...
-- PX4 version: v1.11.0-rc3-2014-g0d4f48a8d2
-- PX4 config file: /home/rusr/px4dev/Firmware/boards/px4/rpi/default.cmake
-- PX4 config: px4_rpi_default
-- PX4 platform: posix
-- PX4 lockstep: disabled
-- cmake build type: RelWithDebInfo
-- Building for code coverage
-- ccache enabled (export CCACHE_DISABLE=1 to disable)
-- build type is RelWithDebInfo
-- PX4 ECL: Very lightweight Estimation & Control Library v1.9.0-rc1-544-g4df0054
-- ROMFS: ROMFS/px4fmu_common
-- Configuring done
-- Generating done
-- Build files have been written to: /home/rusr/px4dev/Firmware/build/px4_rpi_default
[397/808] Building CXX object src/drivers/imu/adis...eFiles/drivers__imu__adis16477.dir/ADIS16477.cpp.o
../../src/drivers/imu/adis16477/ADIS16477.cpp: In member function ‘void ADIS16477::start()’:
../../src/drivers/imu/adis16477/ADIS16477.cpp:275:95: warning: statement has no effect [-Wunused-value]
   px4_arch_gpiosetevent(_drdy_gpio, true, false, true, &ADIS16477::data_ready_interrupt, this);
                                                                                               ^
../../src/drivers/imu/adis16477/ADIS16477.cpp: In member function ‘virtual void ADIS16477::exit_and_cleanup()’:
../../src/drivers/imu/adis16477/ADIS16477.cpp:288:75: warning: statement has no effect [-Wunused-value]
   px4_arch_gpiosetevent(_drdy_gpio, false, false, false, nullptr, nullptr);
                                                                           ^
[808/808] Linking CXX shared library src/examples/dyn_hello/examples__dyn_hello.px4mod

@bkueng The issue is still there...

$ sudo make px4_rpi
-- PX4 version: v1.11.0-rc3-2031-ge57aaaaa5e
-- PX4 config file: /home/rusr/px4dev/Firmware/boards/px4/rpi/default.cmake
-- PX4 config: px4_rpi_default
-- PX4 platform: posix
-- PX4 lockstep: disabled
-- cmake build type: RelWithDebInfo
-- The CXX compiler identification is GNU 7.5.0
-- The C compiler identification is GNU 7.5.0
-- The ASM compiler identification is GNU
-- Found assembler: /usr/bin/arm-linux-gnueabihf-gcc
-- Check for working CXX compiler: /usr/bin/arm-linux-gnueabihf-g++
-- Check for working CXX compiler: /usr/bin/arm-linux-gnueabihf-g++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Check for working C compiler: /usr/bin/arm-linux-gnueabihf-gcc
-- Check for working C compiler: /usr/bin/arm-linux-gnueabihf-gcc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Building for code coverage
-- ccache enabled (export CCACHE_DISABLE=1 to disable)
-- Found PythonInterp: /usr/bin/python3 (found suitable version "3.6.9", minimum required is "3") 
-- build type is RelWithDebInfo
-- PX4 ECL: Very lightweight Estimation & Control Library v1.9.0-rc1-546-gbb950a1
-- ROMFS: ROMFS/px4fmu_common
-- Configuring done
-- Generating done
-- Build files have been written to: /home/rusr/px4dev/Firmware/build/px4_rpi_default
[1/825] git submodule src/drivers/gps/devices
[11/825] git submodule src/lib/ecl
[12/825] git submodule mavlink/include/mavlink/v2.0
[417/825] Building CXX object src/drivers/imu/a...les/drivers__imu__adis16477.dir/ADIS16477.cpp.o
FAILED: src/drivers/imu/adis16477/CMakeFiles/drivers__imu__adis16477.dir/ADIS16477.cpp.o 
/usr/bin/ccache /usr/bin/arm-linux-gnueabihf-g++  -DCONFIG_ARCH_BOARD_PX4_RPI -DMODULE_NAME=\"adis16477\" -DPX4_MAIN=adis16477_app_main -D__CUSTOM_FILE_IO__ -D__PX4_LINUX -D__PX4_POSIX -D__STDC_FORMAT_MACROS -Dnoreturn_function="__attribute__((noreturn))" -I../../boards/px4/rpi/src -I../../platforms/posix/src/px4/common/include -I. -Isrc/lib -I../../platforms/posix/src/px4/generic/generic/include -I../../platforms/common -I../../platforms/common/include -I../../src -I../../src/include -I../../src/lib -I../../src/lib/matrix -I../../src/modules -I../../platforms/posix/include -Iexternal/Install/include -O2 -g -DNDEBUG -fPIC   -g -fdata-sections -ffunction-sections -fomit-frame-pointer -fmerge-all-constants -fno-signed-zeros -fno-trapping-math -freciprocal-math -fno-math-errno -fno-strict-aliasing -fvisibility=hidden -include visibility.h -Wall -Wextra -Werror -Warray-bounds -Wcast-align -Wdisabled-optimization -Wdouble-promotion -Wfatal-errors -Wfloat-equal -Wformat-security -Winit-self -Wlogical-op -Wpointer-arith -Wshadow -Wuninitialized -Wunknown-pragmas -Wunused-variable -Wno-missing-field-initializers -Wno-missing-include-dirs -Wno-unused-parameter -fdiagnostics-color=always -fno-builtin-printf -fno-strength-reduce -Wformat=1 -Wunused-but-set-variable -Wno-format-truncation -fcheck-new -fno-exceptions -fno-rtti -fno-threadsafe-statics -Wreorder -Wno-overloaded-virtual -Wno-cast-align -std=gnu++14 -MD -MT src/drivers/imu/adis16477/CMakeFiles/drivers__imu__adis16477.dir/ADIS16477.cpp.o -MF src/drivers/imu/adis16477/CMakeFiles/drivers__imu__adis16477.dir/ADIS16477.cpp.o.d -o src/drivers/imu/adis16477/CMakeFiles/drivers__imu__adis16477.dir/ADIS16477.cpp.o -c ../../src/drivers/imu/adis16477/ADIS16477.cpp
../../src/drivers/imu/adis16477/ADIS16477.cpp: In member function ‘void ADIS16477::start()’:
../../src/drivers/imu/adis16477/ADIS16477.cpp:275:95: error: statement has no effect [-Werror=unused-value]
   px4_arch_gpiosetevent(_drdy_gpio, true, false, true, &ADIS16477::data_ready_interrupt, this);
                                                                                               ^
compilation terminated due to -Wfatal-errors.
cc1plus: all warnings being treated as errors
[422/825] Building CXX object src/drivers/magne...agnetometer__isentek__ist8310.dir/IST8310.cpp.o
ninja: build stopped: subcommand failed.
Makefile:226: recipe for target 'px4_rpi' failed
make: *** [px4_rpi] Error 1

--- a/platforms/posix/src/px4/common/include/px4_platform/micro_hal.h
+++ b/platforms/posix/src/px4/common/include/px4_platform/micro_hal.h
@@ -32,11 +32,17 @@
  ****************************************************************************/
 #pragma once
 
-#define px4_arch_configgpio(pinset)              -1
-#define px4_arch_unconfiggpio(pinset)
-#define px4_arch_gpioread(pinset)                -1
-#define px4_arch_gpiowrite(pinset, value)        -1
-#define px4_arch_gpiosetevent(pinset,r,f,e,fp,a) -1
+#include <stdint.h>
+#include <stdbool.h>
+
+typedef int (*xcpt_t)(int irq, void *context, void *arg);
+
+static inline int px4_arch_configgpio(uint32_t pinset) { return -1; }
+static inline int px4_arch_unconfiggpio(uint32_t pinset) { return -1; }
+static inline bool px4_arch_gpioread(uint32_t pinset) { return false; }
+static inline void px4_arch_gpiowrite(uint32_t pinset, bool value) { }
+static inline int px4_arch_gpiosetevent(uint32_t pinset, bool risingedge, bool fallingedge,
+					bool event, xcpt_t func, void *arg) { return -1; }
 
 #define px4_udelay(usec) px4_usleep(usec)
 #define px4_mdelay(msec) px4_msleep(msec)

